buildSrc/src/main/kotlin/io/opentelemetry/gradle/OtelVersionClassPlugin.kt

```diff
@@ -0,0 +1,74 @@
+package io.opentelemetry.gradle
+
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.plugins.JavaPlugin
+import org.gradle.api.plugins.JavaPluginExtension
+import org.gradle.kotlin.dsl.the
+import java.io.File
+
+/**
+ * This gradle plugin will define a new task called generateOtelVersionClass.
+ * This task generates a Java source file that contains the project version
+ * as a string constant. The "compileJava" task is updated to depend on
+ * generateOtelVersionClass, and the project source set is updated to
+ * include the new file.
+ */
+class OtelVersionClassPlugin : Plugin<Project> {
+  override fun apply(project: Project) {
+    project.plugins.apply(JavaPlugin::class.java)
+
+    project.task("generateOtelVersionClass") {
+      doLast {
+        writeFile(project)
+      }
+    }
+    addTaskDependency(project)
+    addSourceDependency(project)
+  }
+
+  private fun writeFile(project: Project) {
+    val group = "${project.group}".replace('.', '/')
+    val projectName = project.name.replace('-', '/')
+    val outDir = buildOutDir(project)
+    val filename = "$group/$projectName/internal/OtelVersion.java"
+    val outFile = File(outDir, filename)
+    val packageName = "${project.group}.${project.name.replace('-', '.')}.internal"
+    val classBody = getClassBody("${project.version}", packageName)
+
+    outFile.parentFile.mkdirs()
+    outFile.writeText(classBody)
+  }
+
+  private fun getClassBody(version: String, packageName: String): String {
+    return """
+      package $packageName;
+      
+      /**
+       * Autogenerated class do not edit.
+       */
+       public final class OtelVersion {
+           public static final String VERSION = "$version";
+           private OtelVersion(){}
+       }
+    """.trimIndent()
+  }
+
+  private fun addTaskDependency(project: Project) {
+    project.tasks.getByName("compileJava") {
+      dependsOn("generateOtelVersionClass")
+    }
+  }
+
+  private fun addSourceDependency(project: Project) {
+    val outDir = buildOutDir(project)
+    val java = project.the<JavaPluginExtension>()
+    java.sourceSets.getByName("main").java {
+        srcDir(outDir)
+    }
+  }
+
+  private fun buildOutDir(project: Project): File {
+    return File(project.buildDir, "generated-src/version/java")
```

Comment:
Let's output to the idiomatic place for generated source:
```suggestion
    return File(project.buildDir, "generated/sources/version/java/main")
```